buildSrc/src/main/java/UpdateMapStyleTask.kt

```diff
@@ -0,0 +1,85 @@
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.TaskAction
+import java.io.File
+import java.net.URL
+
+open class UpdateMapStyleTask : DefaultTask() {
+
+    @get:Input var targetDir: String? = null
+
+    @TaskAction fun run() {
+        println("hello")
+
+        val targetDir = targetDir?.let { File(it) } ?: return
+        assert(targetDir.exists())
+
+        val githubDirectoryListingUrl = URL("https://api.github.com/repos/streetcomplete/streetcomplete-mapstyle/contents?ref=jawg")
+        val mapStyleFiles = fetchDirectoryContents(githubDirectoryListingUrl)
+
+        val excludeFilePaths = setOf(
+            ".gitattributes",
+            "LICENSE",
+            "README.md"
+        )
+
+        for (file in mapStyleFiles) {
+            if (file.path in excludeFilePaths) {
+                continue
+            }
+
+            val targetFile = File(targetDir, file.path)
+            val downloadUrl = URL(file.download_url)
+
+            if (file.name.endsWith(".yaml")) {
+                val fileContent = downloadUrl.readText()
+                    .normalizeLineEndings()
+                    .replaceOnlineWithLocalSections()
+
+                targetFile.writeText(fileContent)
+            } else {
+                // e.g. for images: don't try to read/write as UTF-8
+                downloadUrl.openStream().use { it.transferTo(targetFile.outputStream()) }
+            }
+        }
+    }
+
+    private val json = Json { ignoreUnknownKeys = true }
+
+    private fun fetchDirectoryContents(sourceGithubDirectoryUrl: URL): List<GithubDirectoryListingItem> =
+        json.decodeFromString<List<GithubDirectoryListingItem>>(sourceGithubDirectoryUrl.readText())
+            .flatMap { if (it.type == "dir") fetchDirectoryContents(URL(it.url)) else listOf(it) }
+}
+
+private fun String.normalizeLineEndings() = this.replace("\r\n", "\n")
+
+private fun String.replaceOnlineWithLocalSections(): String {
+    val lines = this.split("\n").toMutableList()
+
+    val onlineStartLineIndices = lines.indices.filter { lines[it].trim().startsWith("# for online testing") }
+    val localStartLineIndices = lines.indices.filter { lines[it].trim().startsWith("# for local testing") }
+
+    val onlineRanges = onlineStartLineIndices.zip(localStartLineIndices)
+    onlineRanges.forEach { (start, end) -> assert(start < end) }
```

Comment:
`assert` is kind of deprecated in Java. Not really (I think) but the behavior is not what one would expect. IIRC it can be configured as a JVM flag what should be the behavior of assert and by default it is a no-op in release builds. In any case, assert seems to be meant as a debugging tool rather than to enforce contracts. That's what I learnt ~10 years ago at least, feel free to disabuse me.

In any case, in Kotlin there are `check` (for state checks) and `require` (for ensuring contracts).

Reply:
Ah, I always forget that. I'll change it to `require` then.